File: src/main/kotlin/dev/jefy/connectpro/marketplace/applicaion/command/JobApplicationCommandImpl.java
package dev.jefy.connectpro.marketplace.applicaion.command;


import org.jspecify.annotations.NullMarked;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;

import dev.jefy.connectpro.marketplace.applicaion.dtos.JobApplicationRequest;
import dev.jefy.connectpro.marketplace.applicaion.exception.JobPostNotExistOrValidException;
import dev.jefy.connectpro.marketplace.domain.model.JobApplication;
import dev.jefy.connectpro.marketplace.domain.repository.JobApplicationRepository;
import dev.jefy.connectpro.marketplace.domain.vo.JobApplicationId;
import dev.jefy.connectpro.portfolio.PortfolioClient;
import dev.jefy.connectpro.portfolio.domain.vo.JobPostId;
import dev.jefy.connectpro.shared.application.exceptions.ResourceNotFound;
import dev.jefy.connectpro.user.UserClient;
import dev.jefy.connectpro.user.applicaion.dtos.UserData;
import lombok.RequiredArgsConstructor;

/**
 * @author Jôph Yamba
 */
@NullMarked
@Service
@Transactional
@RequiredArgsConstructor
public class JobApplicationCommandImpl implements JobApplicationCommand {
    private final JobApplicationRepository jobApplicationRepo;
    private final PortfolioClient portfolioClient;
    private final UserClient userClient;
    
    @Override
    public JobApplicationId apply(JobApplicationRequest request) {
        Assert.notNull(request, "JobApplicationRequest must not be null");
        JobPostId jobPostId = JobPostId.of(request.jobPostId());
        if (!portfolioClient.existsAndValidJobPost(jobPostId)) {
            throw new JobPostNotExistOrValidException();
        }
        UserData user = userClient.getCurrentUser();
        JobApplication application = new JobApplication(user.id(), jobPostId, request.motivation());
        jobApplicationRepo.save(application);
        
        return application.getId();
    }

    @Override
    public JobApplicationId accept(JobApplicationId applicationId) {
        JobApplication application = getJobApplication(applicationId);
        application.accept();
        jobApplicationRepo.save(application);
        return application.getId();
    }
    @Override
    public JobApplicationId reject(JobApplicationId applicationId) {
        JobApplication application = getJobApplication(applicationId);
        application.reject();
        jobApplicationRepo.save(application);
        return application.getId();
    }
    
    private JobApplication getJobApplication(JobApplicationId applicationId) {
        Assert.notNull(applicationId, "JobApplicationId must not be null");
        return  jobApplicationRepo.findById(applicationId)
                .orElseThrow(() -> new ResourceNotFound("Job Application with id " + applicationId + " not found"));
    }
}

File: src/main/kotlin/dev/jefy/connectpro/marketplace/application/command/JobApplicationCommandImpl.kt
package dev.jefy.connectpro.marketplace.application.command

import dev.jefy.connectpro.marketplace.application.dtos.JobApplicationRequest
import dev.jefy.connectpro.marketplace.application.exception.JobApplicationNotFound
import dev.jefy.connectpro.marketplace.application.exception.JobPostNotExistOrValidException
import dev.jefy.connectpro.marketplace.domain.JobApplication
import dev.jefy.connectpro.marketplace.domain.repository.JobApplicationRepository
import dev.jefy.connectpro.marketplace.domain.vo.JobApplicationId
import dev.jefy.connectpro.portfolio.PortfolioClient
import dev.jefy.connectpro.portfolio.domain.vo.JobPostId
import dev.jefy.connectpro.shared.infrastructure.annotations.CommandService
import dev.jefy.connectpro.user.UserClient

/**
 * @author Jôph Yamba
 */
@CommandService
class JobApplicationCommandImpl(
    private val jobApplicationRepo: JobApplicationRepository,
    private val portfolioClient: PortfolioClient,
    private val userClient: UserClient
) : JobApplicationCommand {

    override fun apply(request: JobApplicationRequest): JobApplicationId {
        val jobPostId = JobPostId.of(request.jobPostId)
        if (!portfolioClient.existsAndValidJobPost(jobPostId)) {
            throw JobPostNotExistOrValidException()
        }
        val user = userClient.getCurrentUser()
        val application = JobApplication(
            id = JobApplicationId(applicantId = user.id, jobPostId = jobPostId.value), 
            motivation = request.motivation
        )
        jobApplicationRepo.save(application)

        return application.id
    }

    override fun accept(applicationId: JobApplicationId): JobApplicationId {
        val application = getJobApplication(applicationId)
        application.accept()
        jobApplicationRepo.save(application)
        return application.id
    }

    override fun reject(applicationId: JobApplicationId): JobApplicationId {
        val application = getJobApplication(applicationId)
        application.reject()
        jobApplicationRepo.save(application)
        return application.id
    }

    private fun getJobApplication(applicationId: JobApplicationId): JobApplication = jobApplicationRepo
        .findById(applicationId)
        .orElseThrow { JobApplicationNotFound() }
}